core/txpool/pending.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2014 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package txpool
+
+import (
+	"container/heap"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/holiman/uint256"
+)
+
+type Pending interface {
+	// Shift replaces the current best head with the next one from the same account.
+	Shift()
+	// Peek returns the next transaction by price.
+	Peek() (*LazyTransaction, *uint256.Int)
+
+	// Pop removes the best transaction, *not* replacing it with the next one from
+	// the same account. This should be used when a transaction cannot be executed
+	// and hence all subsequent ones should be discarded from the same account.
+	Pop()
+
+	// Empty returns true if the set is empty.
+	Empty() bool
+
+	// Clears the set
+	Clear()
+}
+
+// PendingFilter is a collection of filter rules to allow retrieving a subset
+// of transactions for announcement or mining.
+//
+// Note, the entries here are not arbitrary useful filters, rather each one has
+// a very specific call site in mind and each one can be evaluated very cheaply
+// by the pool implementations. Only add new ones that satisfy those constraints.
+type PendingFilter struct {
+	MinTip  *uint256.Int // Minimum miner tip required to include a transaction
+	BaseFee *uint256.Int // Minimum 1559 basefee needed to include a transaction
+	BlobFee *uint256.Int // Minimum 4844 blobfee needed to include a blob transaction
+
+	OnlyPlainTxs bool // Return only plain EVM transactions (peer-join announces, block space filling)
+	OnlyBlobTxs  bool // Return only blob transactions (block blob-space filling)
+
+	OnlyLocals bool // Return only txs from 'local' addresses.
+	NoLocals   bool // Remove all txs from 'local' addresses
+}
```

Comment:
This thing is getting kind of big. Perhaps, instead of passing around a `PendingFilter`, we should pass around ` *PendingFilter`, which in turn could have `uint256.Int` fields instead of `*uint256.Int`?